### PR TITLE
feature/redis

### PR DESCRIPTION
**Pull request by aish-26-2000**
- initialized redis
- added redisHandler functions
- added rediscache middleware

### CODE REVIEW

**Comment by aish-26-2000** [comment]
Redis caching - workspaces
https://user-images.githubusercontent.com/102617444/237010315-2b8dff33-8f83-41b9-a3d1-935b310204cf.mp4

File: src/modules/utils/redisHandler.ts [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+  await redisClient.set(key, redisValue);
+  const EX = 3600;
+  await redisClient.expire(key, EX);
+};
+
+/**
+ * Deletes a Redis key
+ * @param {string} key
+ */
+export const delRedisKey = async (key: string): Promise<void> => {
```
Comment (sonalidevbud) [inline src/modules/utils/redisHandler.ts:59]: lets add a function to delete a key pattern as well

File: src/modules/redis/redisCache.ts [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+
+export const redisCache = catchAsync(async (req: Request, res: Response, next: NextFunction) => {
+  let key = '';
+  key = req.user !== undefined ? `${req.originalUrl}::${req.user.id}` : `${req.originalUrl}`;
+  const cachedData: string | null = await getFromRedis(key);
+  if (!cachedData) {
+    return next();
+  }
+  const parsedData = cachedData ? JSON.parse(cachedData) : null;
+  const response = paginatedResponseBuilder(httpStatus.OK, 'Success', parsedData);
```
Comment (sonalidevbud) [inline src/modules/redis/redisCache.ts:16]: will this always be paginated response builder or can it be single response builder as well - is there any way of identifying ?

File: src/modules/redis/redisCache.ts [hunk 3]
```diff
@@ -0,0 +1,18 @@
+/* eslint-disable import/prefer-default-export */
+import { Request, Response, NextFunction } from 'express';
+import httpStatus from 'http-status';
+import { catchAsync } from '../utils';
+import { getFromRedis } from '../utils/redisHandler';
+import { paginatedResponseBuilder } from '../utils/responseHandler';
+
+export const redisCache = catchAsync(async (req: Request, res: Response, next: NextFunction) => {
```
Comment (sonalidevbud) [inline src/modules/redis/redisCache.ts:8]: this is a middleware - also it does not require it's own module, we can add this in utils module - we can also keep this function inside redisHandler itself

**Review by sonalidevbud — CHANGES_REQUESTED** [review]

File: src/modules/redis/redisCache.ts (hunk 3 above)
Comment (aish-26-2000) [reply src/modules/redis/redisCache.ts]: done

File: src/modules/utils/redisHandler.ts (hunk 1 above)
Comment (aish-26-2000) [reply src/modules/utils/redisHandler.ts:63]: done

File: src/modules/redis/redisCache.ts (hunk 2 above)
Comment (aish-26-2000) [reply src/modules/redis/redisCache.ts]: image

**Pull request merged**